ath79: add support for GL.iNet GL-MIFI

pesintta opened the pull request:
Add support for the ar71xx supported GL.iNet GL-MIFI to ath79.

Specifications:
 - Atheros AR9331
 - 64 MB of RAM
 - 16 MB of FLASH (SPI NOR)
 - 2x 10/100/1000 Mbps Ethernet
 - 2.4GHz (AR9330), 802.11b/g/n
 - 1x USB 2.0 (vbus driven by GPIO)
 - 4x LED, driven by GPIO
 - 1x button (reset)
 - 1x mini pci-e slot (vcc driven by GPIO)

Flash instructions:

Vendor software is based on openwrt so you can flash the sysupgrade
image via the vendor GUI or using command line sysupgrade utility.
Make sure to not save configuration over reflash as uci settings
differ between versions.

This patch is based on earlier work[1] done by Kyson Lok. Since the
initial mailing-list submission the patch has been modified to comply
with current openwrt naming schemes and dts conventions.

[1] http://lists.openwrt.org/pipermail/openwrt-devel/2018-September/019576.html

Signed-off-by: Antti Seppälä <a.seppala@gmail.com>

pesintta committed:
ath79: add support for GL.iNet GL-MiFi

Add support for the ar71xx supported GL.iNet GL-MiFi to ath79.

Specifications:
 - Atheros AR9331
 - 64 MB of RAM
 - 16 MB of FLASH (SPI NOR)
 - 2x 10/100/1000 Mbps Ethernet
 - 2.4GHz (AR9330), 802.11b/g/n
 - 1x USB 2.0 (vbus driven by GPIO)
 - 4x LED, driven by GPIO
 - 1x button (reset)
 - 1x mini pci-e slot (vcc driven by GPIO)

Flash instructions:

Vendor software is based on openwrt so you can flash the sysupgrade
image via the vendor GUI or using command line sysupgrade utility.
Make sure to not save configuration over reflash as uci settings
differ between versions.

Note on MAC addresses:

Even though the platform is capable to providing separate MAC addresses
to the interfaces vendor firmware does not seem to take advantage of
that. It appears that there is only single unique pre-programmed
address in the art partition and vendor firmware uses that for
every interface (eth0/eth1/wlan0). Similar behaviour has also been
implemented in this patch.

Note on GPIOs:

In vendor firmware the gpio controlling mini pci-e slot is named
3gcontrol while it actually controls power supply to the entire mini
pci-e slot. Therefore a more descriptive name (minipcie) was chosen.
Also during development of this patch it became apparent that the
polarity of the signal is actually active low rather than active high
that can be found in vendor firmware.

Acknowledgements:

This patch is based on earlier work[1] done by Kyson Lok. Since the
initial mailing-list submission the patch has been modified to comply
with current openwrt naming schemes and dts conventions.

[1] http://lists.openwrt.org/pipermail/openwrt-devel/2018-September/019576.html

Signed-off-by: Antti Seppälä <a.seppala@gmail.com>

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+
+/ {
+	compatible = "glinet,gl-mifi", "qca,ar9331";
+	model = "GL.iNet GL-MIFI";
```
Based on GL.iNet's page I'd adjust the capitalization to GL-MiFi ...
https://www.gl-inet.com/products/gl-mifi/

pesintta replied:
Should be addressed in 753db9a

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
```
please use gpio-keys for compatible instead and drop poll-interval.

pesintta replied:
Changed in 3d665a8. Tested that this works too.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+
+	export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
```
This line can be dropped.

pesintta replied:
Done in be34a6e

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+		gpio = <&gpio 6 GPIO_ACTIVE_LOW>;
+	};
+
+	export {
```
"gpio-export"

pesintta replied:
Also changed in be34a6e

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+
+&spi {
+	num-chipselects = <1>;
+	status = "okay";
```
status first, then empty line, then num-chipselects, please.

pesintta replied:
Addressed in 5370c1f

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+				read-only;
+			};
+
+			partition@1 {
```
please use "partition@position" notation, i.e. partition@40000 here and so on for the other partitions.

pesintta replied:
Changed in ae13aaa

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -808,6 +808,16 @@ define Device/glinet_gl-ar750
+define Device/glinet_gl-mifi
+  SOC := ar9331
+  DEVICE_VENDOR := GL.iNET
+  DEVICE_MODEL := GL-MIFI
```
GL-MiFi

pesintta replied:
Changed also in 753db9a

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+		3g {
+			label = "gl-mifi:green:3g";
+			gpios = <&gpio 0 GPIO_ACTIVE_HIGH>;
+		};
```
This has been called "gl-mifi:green:net" beforehand. One should add an entry in generic/base-files/etc/uci-defaults/04_led_migration for that led. Despite, on the web page it's labelled "3g/4g". Maybe we can still improve that name ...

pesintta replied:
Changed the led name to "3g4g" in commit 521fb28 where also simple migration entry was added. Let me know if a better name comes along...

adrianschmutzler replied:
3g4g is also my favorite of all the suboptimal ideas I had ...

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+		#size-cells = <0>;
+
+		minipcie {
+			gpio-export,name = "minipcie";
```
In ar71xx this has been set up as "3gcontrol", and with active_high...

pesintta replied:
In my opinion the choice of 3gcontrol as name is rather poor as the gpio actually controls the minipci-e slot vcc where various different modules could be (for example I use LTE module instead of 3g).
Also I double checked that the module gets powered on (= is active) when the signal is low so I'm inclined to believe that ar71xx does not do the right thing here.

How do you want to proceed with this one?

adrianschmutzler replied:
I'm a fan of using the ar71xx->ath79 process for improvements, so if you know the truth I'm all in for changing stuff if it was wrong/worse in ar71xx.

adrianschmutzler replied:
Consider sending a separate patch for ar71xx as well to fix the active_high vs. active_low there ...

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+};
+
+&eth0 {
+	status = "okay";
```
I would have added a
```
	gmac-config {
		device = <&gmac>;

		switch-phy-addr-swap = <0>;
		switch-phy-swap = <0>;
	};
```
here due to the `ath79_setup_ar933x_phy4_switch(false, false);` in ar71xx (see ar9331_glinet_64xx.dtsi). But I'm not really sure whether that's required if both options are set to false.

pesintta replied:
In my tests I did not find any reason to have this block but the function may change status of some configuration bits so maybe it is better to be added to mimic ar71xx behavior. Therefore I added the dts block in 8ac0c63.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <1>;
```
This increment is not used in ar71xx, please drop it unless you find it's used in vendor firmware.

pesintta replied:
Double-checked behaviour with vendor firmware and sadly it does not do such a thing so I dropped the increment in 
02d892b.

Hopefully openwrt does not run into any issues with having every interface use the exact same mac-address.

adrianschmutzler replied:
Unfortunately vendors quite frequently do this, and I don't think it's wise to deviate (at least) for _default_ setup.
I have already seen consecutive addresses on devices in a real network, so better be safe here.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,152 @@
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x0>;
```
Have you checked whether there is a suitable address in 0x1002, i.e. one that will be read automatically from caldata?

pesintta replied:
Based on this tip I examined the contents of the art partition and it turns out there really is mac address stored in 0x1002 (sadly it is not different from the main mac at 0x0) so I guess the separate mtd-mac-address line can be dropped from here which I did in d15f529.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -146,6 +146,10 @@ glinet,gl-ar300m-lite)
 	;;
+glinet,gl-mifi)
+	ucidef_set_led_netdev "lan" "LAN" "$boardname:green:lan" "eth0"
+	ucidef_set_led_netdev "wan" "WAN" "$boardname:green:wan" "eth1"
```
For correct link detection, typically one of the interfaces needs to be set up with ucidef_set_led_switch. I'm never sure which, though. (Take this as optional improvement.)

pesintta replied:
After some fiddling I found out the switch offset for the external lan port and changed led to use that in 21e0a0d.

pesintta commented:
Thank you for the review @adrianschmutzler!

I've pushed bunch of commits to address most of the comments and improvement suggestions. I'll keep the commits separate for easier review and will squash them together once everything looks good enough to proceed.

adrianschmutzler commented:
> I've pushed bunch of commits to address most of the comments and improvement suggestions. I'll keep the commits separate for easier review and will squash them together once everything looks good enough to proceed.

For me, review is easier with a single, proper commit. With multiple commits, I will use the overall changes display anyway (unless there is a reason for multiple commits, and they would go into master separately).

So, you might speed things up by squashing right away.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,155 @@
+
+&gpio {
+	status = "okay";
+};
```
gpio node can be dropped due to a recent patch.

pesintta replied:
Quite recent indeed :)
Dropped the node as requested.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_glinet_gl-mifi.dts:
```diff
@@ -0,0 +1,155 @@
+&eth0 {
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
```
I'd prefer an empty line before gmac-config.

pesintta replied:
Yeah, this is included this in the latest version.

adrianschmutzler commented:
Content looks good, and the commit message from the first commit seems to be fine on a quick look as well.

Please add a note to the commit message:
- ... that you have verified the MAC addresses with vendor firmware, and they are the same for lan/wan/wifi there (otherwise, in two years nobody will know whether you actually checked or not).
- ... that you have verified the active-high vs. active-low for the gpio-export, and maybe the reason for "renaming" it.

Wrap it up and you might get a quick merge here. :-)

pesintta commented:
> Content looks good, and the commit message from the first commit seems to be fine on a quick look as well.
> 
> Please add a note to the commit message:
> 
> * ... that you have verified the MAC addresses with vendor firmware, and they are the same for lan/wan/wifi there (otherwise, in two years nobody will know whether you actually checked or not).
> * ... that you have verified the active-high vs. active-low for the gpio-export, and maybe the reason for "renaming" it.
> 
> Wrap it up and you might get a quick merge here. :-)

Added minor edits according to review comments above.
Squashed commits, updated commit message with additional notes and pushed.

adrianschmutzler commented:
Thanks, pulled into my staging tree.

xback commented:
I just compiled and flashed this.

Ethernet does not work correctly.

```
[   47.617453] eth0: link up (1000Mbps/Full duplex)
[   47.637123] br-lan: port 1(eth0) entered blocking state
[   47.640904] br-lan: port 1(eth0) entered disabled state
[   47.646619] device eth0 entered promiscuous mode
[   47.695978] br-lan: port 1(eth0) entered blocking state
[   47.699847] br-lan: port 1(eth0) entered forwarding state
[   49.858890] eth1: link up (100Mbps/Full duplex)
```
eth0 is the WAN port, which is only 100Mbps capable, not 1Gbit
eth1 is the LAN port, which is 1Gbit capable.

Above shows that both ports are coming up .. while there is only a cable attached to eth0

Here is the log from 19.07 with also only a cable attached to eth0

```
[   53.869999] br-lan: port 1(eth1) entered blocking state
[   53.881249] br-lan: port 1(eth1) entered disabled state
[   53.886875] device eth1 entered promiscuous mode
[   53.938109] br-lan: port 1(eth1) entered blocking state
[   53.941900] br-lan: port 1(eth1) entered forwarding state
[   54.056817] br-wan: port 1(eth0) entered blocking state
[   54.060606] br-wan: port 1(eth0) entered disabled state
[   54.066338] device eth0 entered promiscuous mode
[   54.896030] br-lan: port 1(eth1) entered disabled state
[   58.176914] eth0: link up (100Mbps/Full duplex)
[   58.180047] br-wan: port 1(eth0) entered blocking state
[   58.185212] br-wan: port 1(eth0) entered forwarding state
[   68.233266] done.
```

pesintta commented:
> eth0 is the WAN port, which is only 100Mbps capable, not 1Gbit
> eth1 is the LAN port, which is 1Gbit capable.

Unless I'm missing something my unit has eth1 labeled as WAN. Here are messages printed when cable is connected there:
```
[ 1387.730227] eth1: link up (100Mbps/Full duplex)
[ 1387.733657] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
```

Here are the messages about eth0 when device is booted and cable is connected
```
[   34.905309] eth0: link up (1000Mbps/Full duplex)
[   34.909496] br-lan: port 1(eth0) entered blocking state
[   34.913780] br-lan: port 1(eth0) entered disabled state
[   34.919445] device eth0 entered promiscuous mode
[   35.904255] br-lan: port 1(eth0) entered blocking state
[   35.908038] br-lan: port 1(eth0) entered forwarding state
[   35.945066] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
```

In other words on my unit the ethernet ports work correctly.

Do you perhaps have some previous configuration that would flip LAN & WAN ports or did you flash it with sysupgrade -n?

xback commented:
@pesintta 
I just compiled both latest states from 19.07 and master repo's

In 19.07
Eth0 matches label WAN as printed on the device (links up to 100mbit)
Eth1 matches label LAN as printed on the device (links up to 1gbit)

Both ports work correctly when plugging and unplugging etherney cables.

On master:
Eth0 now shows itself as 1gbit
Eth1 now shows itself as 100mbit

Eth1 linkstate does not respond when the cable is plugged/unplugged.
The port is always up.

See the dmesg prints in my previous post.

I cannot ping either interface over IP.

Both images were flashed using the bootloader recovery webui to ensure the full flash is cleared.

Any idea?

pesintta commented:
I'm sorry but I'm running out of ideas. For me the latest image from master works perfectly. Can you post full dmesg from both versions so I can compare it with mine?

It's almost as if we are working with different hardware or something...

xback commented:
Idd strange.
Thanks for checking.

Ill provide the full dmesg tomorrow from 2 new builds from both branches.

Ill also post some pictures of the device.

xback commented:
@pesintta 

I retested on a different glmifi board.
Ethernet does not seem to be stuck now (hw issue?)

I can confirm that both ethernet ports are switched.
The bootlog also shows some different behaviour in boot order:

19.07
WAN = eth0  (10/100)
LAN = eth1  (10/100/1000)
[bootlog](https://pastebin.com/raw/68x51qmx)

master
WAN = eth1  (10/100)
LAN = eth0  (10/100/1000)
[bootlog](https://pastebin.com/raw/7QudRS3D)

Hope this helps.

Thanks

pesintta commented:
> 
> 
> @pesintta
> 
> I retested on a different (identical) glmifi board.
> Ethernet does not seem to be stuck now and both ports can transceive data (hw issue?)
> 

Ok, this is good 👍 

> I can confirm that both ethernet ports are switched.
> 

This is apparently "not a bug" but just the way ath79 target works which is different from ar71xx. The probing order of ethernet devices is fixed in ath79 and there's no way to change that in DTS.
See discussion thread from roughly a year ago where end-result was that migration for such situation would not be developed: http://lists.openwrt.org/pipermail/openwrt-devel/2020-January/027085.html

I guess the best we could do is to add GL-MiFi to the list of devices without full migration available here: https://openwrt.org/docs/guide-user/installation/ar71xx.to.ath79#devices_with_known_config_changes_without_migration_available

adschm commented:
One can use stuff like this to force the "old" order:

https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_openmesh_om2p-v4.dtsi#L143-L146

However, on a device already merged into ath79 one would just have the choice between breaking upgrade between newer or between new and older versions ...
